core/src/main/java/io/grpc/internal/DnsNameResolver.java

```diff
@@ -226,4 +234,153 @@ public final synchronized void shutdown() {
   final int getPort() {
     return port;
   }
+
+  private DelegateResolver pickDelegateResolver() {
+    JdkResolver jdkResolver = new JdkResolver();
+    if (isJndiAvailable) {
+      return new CompositeResolver(jdkResolver, new JndiResolver());
+    }
+    return jdkResolver;
+  }
+
+  /**
+   * Forces the resolver.  This should only be used by testing code.
+   */
+  @VisibleForTesting
+  void setDelegateResolver(DelegateResolver delegateResolver) {
+    this.delegateResolver = delegateResolver;
+  }
+
+  /**
+   * Returns whether the JNDI DNS resolver is available.  This is accomplished by looking up a
+   * particular class.  It is believed to be the default (only?) DNS resolver that will actually be
+   * used.  It is provided by the OpenJDK, and likely Android.  Actual resolution will be done by
+   * using a service provider when a hostname query is present, so the {@code DnsContextFactory}
+   * may not actually be used to perform the query.  This is believed to be "okay."
+   */
+  @VisibleForTesting
+  @SuppressWarnings("LiteralClassName")
+  static boolean jndiAvailable() {
+    try {
+      Class.forName("com.sun.jndi.dns.DnsContextFactory");
+    } catch (ClassNotFoundException e) {
+      logger.log(Level.FINE, "Unable to find JNDI DNS resolver, skipping", e);
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Common interface between the delegate resolvers used by DnsNameResolver.
+   */
+  @VisibleForTesting
+  abstract static class DelegateResolver {
+    abstract ResolutionResults resolve(String host) throws Exception;
+  }
+
+  /**
+   * Describes the results from a DNS query.
+   */
+  @VisibleForTesting
+  static final class ResolutionResults {
+    final List<InetAddress> addresses;
+    final List<String> txtRecords;
+
+    ResolutionResults(List<InetAddress> addresses, List<String> txtRecords) {
+      this.addresses = Collections.unmodifiableList(checkNotNull(addresses, "addresses"));
+      this.txtRecords = Collections.unmodifiableList(checkNotNull(txtRecords, "txtRecords"));
+    }
+  }
+
+  /**
+   * A composite DNS resolver that uses both the JDK and JNDI resolvers as delegate.  It is
+   * expected that two DNS queries will be executed, with the second one being from JNDI.
+   */
+  @VisibleForTesting
+  static final class CompositeResolver extends DelegateResolver {
+
+    private final DelegateResolver jdkResovler;
+    @Nullable private final DelegateResolver jndiResovler;
```

Comment:
Done.

I feel like you are pushing into premature simplification.  This code was very easy to modify to accept results from other, future resolvers, but now it is going to be harder.